### PR TITLE
feat(UI): character preview in traits tab on character creation

### PR DESCRIPTION
**Pull request by 4nonch**
### Required

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [x] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [x] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [x] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

### Optional

- [x] This is a PR that modifies build process or code organization.
  - [ ] Please document the changes in the appropriate location in the `doc/` folder.
  - [ ] If documentation for this feature or process does not exist, please write it.
  - [ ] If the change alters versions of software required to build or work with the game, please document it.

## Purpose of change
Character creation menu offers no way to see how your character will look in game with enabled tileset. There could be many traits that influence on character appearance. With mods there could be dozen of different hair/skin/eyes and other types of traits. And you pick that stuff only by guessing. If you don't like the result - you'll have to restart the game and try your luck again or manually edit traits ingame via debugger menu.

I'm offering to make character preview feature that will help players to immediately track character appearance changes and pick the traits they find proper.

## Describe the solution
- "cata_tiles::tile_type_search" method were created from "cata_tiles::draw_from_id_string" for possible reusability later and reduce cata_tiles::draw_from_id_string complexity (actually thought i'll need that part of code to get tile but there were no big need at the end. I'll revert the change if it need to)
- "cata_tiles::draw_from_id_string" signature changed. "bool as_independent_entity = false" added to specify the need of rendering single tile without considering game map context and player position on a screen
- "cata_tiles::draw_entity_with_overlays" signature changed.  "bool as_independent_entity = false" added to bypass it to "cata_tiles::draw_from_id_string"
- "void cata_tiles::display_character( const Character &ch, const point &p )" public method were made to display character tile on some point of the screen
- "character_preview_window" structure were created to handle character window preview
- "termx_to_pixel_value" and "termy_to_pixel_value" were added to get pixels value of single terminal value (like that 1 termy unit is 8 pixels length)
- Changes in "avatar::create" and "tab_direction set_traits" in newcharacter.cpp to integrate solution

## Describe alternatives you've considered
The character preview window is probably too big. I thought to add additional information to it to make it feel not so empty or to change the design but it requires some conversation. 

## Testing
Was able to build on windows 10 under MSYS (cmake) with help of [build guide](https://docs.cataclysmbn.org/en/dev/guides/building/cmake/#windows-environment-msys2)
Went through some manual tests in character creation menu and in game. Not noticed any bugs (for now).

## Additional context

See a video with a showcase below:
[Showcase](https://files.catbox.moe/pbpub6.mp4)

### CODE REVIEW

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -3181,6 +3215,67 @@ std::string points_left::to_string()
...
+#endif
+}
+
+void character_preview_window::zoom_out()
+{
+#if defined(TILES)
+    zoom = zoom / 2;
+    if( zoom < MIN_ZOOM ) {
+        zoom = MAX_ZOOM;
+    }
```
Comment (scarf005) [inline src/newcharacter.cpp:3260]:
```suggestion
    zoom = std::min( zoom / 2, MAX_ZOOM );
```

File: src/newcharacter.h [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 #include <string>
 
+#include "cursesport.h"
```
Comment (scarf005) [inline src/newcharacter.h:7]:
```suggestion
#include "cursesdef.h"
#include "cursesport.h"
```

including cursesdef.h is needed to silence errors in error in clang++-17 (idk why it did build on clang++-17 but warns it in clangd)
```
Field has incomplete type 'catacurses::window' (fix available)clang(field_incomplete_or_sizeless)
cursesport.h(16, 7): Forward declaration of 'catacurses::window'
```

also forward declaration of `avatar` is needed to fix

```
Unknown type name 'avatar' (fix available)clang(unknown_typename)
```

```c++
class avatar;
```

File: src/cata_tiles.cpp [hunk 3]
```diff
@@ -644,6 +644,198 @@ void cata_tiles::set_draw_scale( int scale )
     tile_ratioy = ( static_cast<float>( tile_height ) / static_cast<float>( fontheight ) );
 }
 
+std::optional<std::tuple<const tile_type *, std::string>> cata_tiles::tile_type_search(
+            const std::string &id, TILE_CATEGORY category,
+            const std::string &subcategory, int subtile, int rota
+        )
```
Comment (scarf005) [inline src/cata_tiles.cpp:650]: might be a better idea to use structs?

```c++
struct tile_type {
	tile_type* tt;
	std::string found_id;
};
```

```suggestion
auto cata_tiles::tile_type_search(
            const std::string &id, TILE_CATEGORY category,
            const std::string &subcategory, int subtile, int rota
        ) -> std::optional<tile_type>
```

**Review by scarf005 — CHANGES_REQUESTED** [review]
- functionality looks good.
- could also include it in debug mutation window in later PRs.
- code (tbf, entirity of rendering functions) could use a bit of refactoring in later PRs.
- windows build is failing for unknown reasons, will have to pass all tests before merging.

File: src/newcharacter.cpp (hunk 1 above)
Comment (4nonch) [reply src/newcharacter.cpp:3260]: But with min() zoom out will not be able to reset at MAX_ZOOM when it reach MIN_ZOOM or bellow (like default zoom in-game behavior)

File: src/cata_tiles.cpp (hunk 3 above)
Comment (4nonch) [reply src/cata_tiles.cpp]: Added struct: https://github.com/cataclysmbnteam/Cataclysm-BN/pull/5031/files#diff-567dea2474d5637f37f5066aa4e52c19f755ef11f674795c59511c01ef4a4cf8R349

File: src/newcharacter.cpp (hunk 1 above)
Comment (scarf005) [reply src/newcharacter.cpp:3260]: oops, overlooked that

**Comment by scarf005** [comment]
clang-tidy has lint issues, and windows test fails. on discord it was suggested that tile_ptr null ptr reference was the cause: there should be a variable named smthn like `is_test` to determine whether it's running on test mode. maybe we could use that.

**Review by scarf005 — APPROVED** [review]
LGTM

**Comment by chaosvolt** [comment]
Hmm. I didn't really eyeball this until just now. I feel like it's a bit silly that the preview doesn't show any of the character's items on them, and a giant screen that takes up half the game window to barely use that space for a single space is both really not optimal, and at bare minimum is something that really should have an option to disable.

I can also 100% see potential for complaints about not being able to disable this given people might be playing this game in situations where a naked paper doll unavoidably taking up about a quarter of the game window is going to be harder to explain to one's boss than the character sprite would be during normal gameplay...

**Comment by 4nonch** [comment]
> preview doesn't show any of the character's items on them

Well, character preview is intended to display changing character appearance based of traits. With clothes it gets less obvious that character changed somehow because it hides the skin. If you insist - i'll try to add a feature to toggle clothes on hotkey
> giant screen

Tried to make it scalable. Also you can scale character tile itself by pressing Z / Shift+Z to pick appropriate one. If i'll make small screen with character in the middle of great nothing it will look even weirder. If you have suggestion how to make it better please share, will try to implement.
> should have an option to disable.

Will implement that with "disabled" by default.
>PR broke curses
Oops. The idea that curses build incapable of referencing to cata_cursesport is quite strange. Will look what i could do about it.

**Comment by scarf005** [comment]
maybe we could set the panel size to 1/4 and put it on rightmost, and enable it by default (but configurable)

**Pull request merged**